gnupg 2.4.0

fanquake opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

SMillerDev commented:
What is the upstream support policy for 2.3?

p-linnane commented:
From my research [2.3 and 2.4 are the same branch](https://gnupg.org/download/):
>For GnuPG 2.4 (2.2) we consider 2.3.0 (2.1.0) as the birth of the branch.

In the [release announcement](https://lists.gnupg.org/pipermail/gnupg-announce/2022q4/000477.html) they state:
>- Version 2.4 is the current stable version with a lot of new features
  compared to 2.2.  This announcement is about the latest release of
  this series; the previous release was 2.3.8.

So we should just move forward with a standard version bump and not create `gnupg@2.3`.

chenrui333 commented:
```
  mkdir: setup.scm: File exists
  mkdir: setup.scm: Not a directory
```

fanquake commented:
>   mkdir: setup.scm: File exists
> mkdir: setup.scm: Not a directory

Updated to use the patch I added in https://github.com/Homebrew/formula-patches/pull/877, which should fix this failure.

SMillerDev commented on Formula/gnupg.rb:
```diff
@@ -38,12 +38,12 @@ class Gnupg < Formula
   patch do
-    url "https://dev.gnupg.org/rG7011286ce6e1fb56c2989fdafbd11b931c489faa?diff=1"
-    sha256 "407011d4ae9799f50008b431df60cd5b781dca0f572e956fd46245aa209af7e8"
+    url "https://raw.githubusercontent.com/Homebrew/formula-patches/6ce59559708c7f6acd025b5d11c536016c3129e5/gnupg/2.4.0-setup-scm.patch"
```
Why not just from upstream?
```suggestion
    url "https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=patch;h=e89d57a2cb10bd04d266165015f159be2ab48984"
```

fanquake replied:
I wasn't sure which approach was preferred. Have switched to that (will drop the formula-patch if this is merged).

SMillerDev commented on Formula/gnupg.rb:
```diff
@@ -38,12 +38,12 @@ class Gnupg < Formula
+    sha256 "3316f59ba64e698187832471585ecfcb1756460954fcc69b336e1e9860872713"
   end
 
   def install
```
Please replace `"--disable-dependency-tracking"` and `--prefix` in `./configure` with `*std_configure_args`

fanquake replied:
Replaced.

SMillerDev commented:
```
  =================== 
  3 tests run, 3 succeeded, 0 failed, 0 failed expectedly, 0 succeeded unexpectedly, 0 skipped. 
  =================== 
  Making check in gpgme
  mkdir: setup.scm: File exists
  mkdir: setup.scm: Not a directory
  make[2]: *** [xcheck] Error 1
  make[1]: *** [check-recursive] Error 1
  make: *** [check-recursive] Error 1
```

SMillerDev commented:
And the patch did apply, so clearly that's not the complete fix.
```
  ==> Applying gnupg.git-e89d57a2cb10bd04d266165015f159be2ab48984.patch
  patch -g 0 -f -p1 -i /private/tmp/gnupg--patch-20221229-84626-bl6246/gnupg.git-e89d57a2cb10bd04d266165015f159be2ab48984.patch
  patching file tests/gpgme/Makefile.am
  patching file tests/gpgme/all-tests.scm
```

cho-m commented:
If upstream prefers not building in-source-tree, we should probably just do the same.

Can try something similar to GCC formulae and use a `build` directory.

fanquake commented:
Little time to follow up here at the moment, so happy for someone to experiment / take-over.

cho-m commented:
Will work on in #120045
